### PR TITLE
applications: asset_tracker: Add battery monitor

### PR DESCRIPTION
**Pull request by hmhalvorsen**
A battery charge monitor is added to the asset tracker in order
to govern the current charge level of the device.

Signed-off-by: Henrik Malvik Halvorsen <henrik.halvorsen@nordicsemi.no>

### CODE REVIEW

**Comment by lemrey** [comment]
I think it could be a good idea to try and upstream part of this.

**Comment by metc** [comment]
There is a simple interface to read battery voltage from Modem using the AT command `AT%XVBAT`.
Is it something that could be added as well? AT commands are available for all boards that use Modem. No external sensor is required.

File: applications/asset_tracker/src/main.c [hunk 1]
```diff
@@ -348,6 +356,41 @@ static void cloud_cmd_handler(struct cloud_command *cmd)
...
+	    bat_level_acceptable == true) {
+		ui_led_set_color(CONFIG_BATTERY_MONITOR_RED_VALUE,
+				 0,
+				 0,
+				 UI_LED_ON_PERIOD_CRITICAL,
+				 UI_LED_OFF_PERIOD_CRITICAL);
+		bat_level_acceptable = false;
+	} else if (bat_charge >= CONFIG_BATTERY_MONITOR_THRESHOLD &&
+	   bat_level_acceptable == false) {
+		ui_led_set_color(0,
```
Comment (jhn-nordic) [inline applications/asset_tracker/src/main.c]: This breaks current pattern state set by ui_led_set_pattern() and is a bit confusing.
you could do ui_led_set_patttern(ui_led_get_pattern());

File: applications/asset_tracker/Kconfig [hunk 2]
```diff
@@ -145,6 +145,43 @@ config FLIP_POLL_INTERVAL
 		Flip polling interval in milliseconds.
 endif
 
+config BATTERY_MONITOR
+	bool "Use battery charge monitoring"
+	help
+	    Poll the current battery charge of the device and
```
Comment (rlubos) [inline applications/asset_tracker/Kconfig]: Other help entries use different indentation, let's be consistent here.

File: applications/asset_tracker/Kconfig [hunk 3]
```diff
@@ -145,6 +145,43 @@ config FLIP_POLL_INTERVAL
...
+	help
+	    Battery monitor poll interval in milliseconds.
+
+config BATTERY_MONITOR_RED_VALUE
+	int "RED light intensity for low charge percentage"
+	default 127
+	help
+	    The light intensity of the red LED if
+	    the battery charge drops below a threshold value.
+	    Range 0 - 255 (Lowest to highest)
```
Comment (rlubos) [inline applications/asset_tracker/Kconfig]: You can specify the range at Kconfig level with `range 0 255`, see https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/c48ba0f4eaad5111af429ed594acd9a306c2f7b8/subsys/bluetooth/Kconfig.pool#L12 for example use.

Same below.

File: applications/asset_tracker/src/main.c [hunk 4]
```diff
@@ -31,6 +32,12 @@
 #define FLIP_POLL_INTERVAL		0
 #endif
 
+#if CONFIG_BATTERY_MONITOR
+#define BATTERY_MONITOR_INTERVAL	K_MSEC(CONFIG_BATTERY_MONITOR_INTERVAL)
+#else
+#define BATTERY_MONITOR_INTERVAL		0
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: Indentation.

File: applications/asset_tracker/src/main.c [hunk 5]
```diff
@@ -608,6 +655,9 @@ static void work_init(void)
 	k_work_init(&send_flip_data_work, send_flip_data_work_fn);
 	k_delayed_work_init(&flip_poll_work, flip_send);
 	k_delayed_work_init(&long_press_button_work, accelerometer_calibrate);
+#if CONFIG_BATTERY_MONITOR
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: Why not being consistent and use `IS_ENABLED` macro here as well? You don't need to conditionally include `battery_monitor_check` sources, the linker should handle this.

File: applications/asset_tracker/src/main.c [hunk 6]
```diff
@@ -494,7 +537,6 @@ static void sensor_data_send(struct cloud_channel_data *data)
 	if (err) {
 		printk("Unable to encode cloud data: %d\n", err);
 	}
-	err = cloud_encode_data(data, &output);
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: Is this change related/intentional?

File: applications/asset_tracker/src/main.c [hunk 7]
```diff
@@ -348,6 +356,41 @@ static void cloud_cmd_handler(struct cloud_command *cmd)
...
+static void battery_monitor_check(struct k_work *work)
+{
+	static u8_t bat_charge;
+	static bool bat_level_acceptable = true;
+
+	battery_monitor_read(&bat_charge);
+
+	if (bat_charge < CONFIG_BATTERY_MONITOR_THRESHOLD &&
+	    bat_level_acceptable == true) {
+		ui_led_set_color(CONFIG_BATTERY_MONITOR_RED_VALUE,
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: I find it confusing to have `set_color` function accept not only the color, but also the on/off period.

I saw that there was a `UI_BATTERY_LEVEL_LOW` led pattern specified in this PR. Why don't we use it here instead? You could:
1) When the battery level drops below the threshold, store the current pattern and apply the `UI_BATTERY_LEVEL_LOW`.
2) Once battery is connected/charged, restore the previously stored pattern, as suggested by @jhn-nordic.

Of course, you would need to handle situations, when the pattern is being changed while the battery level is low. A simple wrapper over `ui_led_set_patttern` should handle that though.

File: include/modem_info.h [hunk 8]
```diff
@@ -95,6 +95,7 @@ struct device_param {
 	const char *board; /**< Board version. */
 	const char *app_version; /**< Application version. */
 	const char *app_name; /**< Application name. */
+	double bat_charge; /**< The current battery charge, in percentage. */
```
Comment (rlubos) [inline include/modem_info.h]: Do we need a floating point precision here? Isn't an integer in range 0 - 100 accurate enough (we deal with percents here, right?).

File: lib/modem_info/modem_info_json.c [hunk 9]
```diff
@@ -171,6 +171,10 @@ static int device_data_add(struct device_param *device, cJSON *json_obj)
 	total_len += json_add_str(json_obj, "appVersion", device->app_version);
 	total_len += json_add_str(json_obj, "appName", device->app_name);
 
+	if (device->bat_charge < 0xFF) {
```
Comment (rlubos) [inline lib/modem_info/modem_info_json.c]: I'm slightly confused here - states that this value indicates the percentage battery charge level. Why `255` then? I hope we do not expect the battery to be charged for more than 100% :D

**Comment by martiles** [comment]
Are you aware of the [%VBAT](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fref_at_commands%2FREF%2Fat_commands%2Fmob_termination_ctrl_status%2Fxvbat_set.html&resultof=%22%76%62%61%74%22%20) AT-Command, which reads the battery voltage in mV for the nRF9160?
The AT command works so that if the modem is **not activated,** the command does an instant reading of VBAT and if the modem **is activated,** then the AT command responds back the last VBAT value that was taken during the time when RF was not transmitting

(_It has 5 bits for decimals of Volt => that means when the result is in mV, the granularity of results is 1000mV/32 = 31,25mV or the results are multiplies of 31,25mV.

So you can't get exactly 3100mV as a result, but you would get 3125 or 3093._)

**Comment by ojousima** [comment]
Can the %VBAT AT-command be synchronized to take a read right after RF transmission or other power hungry activity? Especially when batteries are drained or cold the voltage droop can be pretty significant and it's more relevant to know if the battery has enough energy to keep transmitting rather than to keep the device on in idle.

**Comment by hmhalvorsen** [comment]
I am fully aware of the %XVBAT-command, and I've used it in previous work of mine, please take a look at the modem info library.

The purpose of this PR is to add a battery module that is able to get an accurate reading of the battery charge if a fuel gauge is present. This is the case for the Thingy:91 for example, where the supplied voltage to the SiP is not always the voltage of the battery, but USB as well if it is present. 

Expanding this module further and making it more general is outside of the scope of my work and this PR.

**Comment by hmhalvorsen** [comment]
I've decided that this module needs work and will therefore close this PR until said work is done.
